README
Wx-Perl-PluggableTextCtrl version 0.01
======================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2017 by Hans Jeuken

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Wx::Perl::PluggableTextCtrl.

0.01  Sat Nov 18 21:42:16 2017
	- original version; created by h2xs 1.23 with options
		-X Wx::Perl::PluggableTextCtrl



lib/Wx/Perl/PluggableTextCtrl/UndoRedo.pm
package Wx::Perl::PluggableTextCtrl::UndoRedo;


use strict;
use warnings;
use Carp;

use base qw( Wx::Perl::PluggableTextCtrl::BasePlugin );
use Wx::Event qw( EVT_CHAR );

my $debug = 1;

if ($debug) {
   use Data::Dumper;
}

sub new {
   my $class = shift;
   my $self = $class->SUPER::new(@_);

   $self->{BUFFER} = "";
   $self->{BUFFERMODE} = '';
   $self->{BUFFERNEXTCALL} = 0;
   $self->{BUFSTART} = 0;
   $self->{FLUSHKEYS} = [9, 13, 32];
   $self->{LASTCALL} = "";
   $self->{REPLACEBUFFER} = "";
   $self->{REDOSTACK} = [];
   $self->{UNDOSTACK} = [];
   my $tc = $self->TxtCtrl;
   $self->Require('KeyEchoes');
   $tc->Listening(1);
   unless($self->Broadcast('echo', 25, 'redo')) { carp "could not assign Ctrl-y to redo\n" }
   unless($self->Broadcast('echo', 26, 'undo')) { carp "could not assign Ctrl-z to undo\n" }
   $self->Commands(
      'backspace' => \&BufferModeBs,
      'cancel' => \&CancelCall,
      'canundo' => \&CanUndo,
      'canredo' => \&CanRedo,
      'clear' =>\&Clear,
      'delete' => \&BufferModeDel,
      'replace' => \&RecordReplace,
      'redo' => \&Redo,
      'undo' => \&Undo,
      'remove' => \&RecordRemove,
      'write' => \&RecordWrite,
      'writeable' => \&BufferModeWr,
   );
   if ($debug) { print "dumping commands "; print Dumper $self->{COMMANDS} }

   return $self;
}

sub Buffer {
   my $self = shift;
   if (@_) { $self->{BUFFER} = shift; }
   return $self->{BUFFER};
}

sub BufferCall { #just invented a new verb here: to buffer, store this call in the character buffer.

   my $self = shift; 
   
   if ($self->BufferNextCall) {
      if ($debug) { print "BufferCall called\n"; }
      my $call = shift; my $index = shift; my $char = shift;
      my $otxt = undef;
      if ($call eq 'replace') { $otxt = shift }
      my $sel = shift; my $pos = shift;

      unless ($sel or (length($char) ne 1)) {
         my $mode = $self->BufferMode;

         if ($mode eq 'backspace') {
            if ($call eq 'remove') {
               if ($debug) { print "performing backspace: '$char'\n" }
               $self->BufferNextCall(0);
               my $buf = $self->Buffer;
               $self->Buffer($char . $buf);
               $self->BufStart($index);
               $self->LastCall($call);
               return 1
            } else {
               carp "unexpected call '$call', it should have been 'remove'\n";
            }
         } elsif ($mode eq 'delete') {
            if ($call eq 'remove') {
               if ($debug) { print "performing delete: '$char'\n" }
               $self->BufferNextCall(0);
               my $buf = $self->Buffer;
               $self->Buffer($buf . $char);
               $self->BufStart($index);
               $self->LastCall($call);
               return 1
            } else {
               carp "unexpected call '$call', it should have been 'remove'\n";
            }

         } elsif ($mode eq 'writeable') {
            if (($call eq 'write') or ($call eq 'replace')) {
               if ($debug) { print "performing writeable: '$char'\n" }
               $self->BufferNextCall(0);
               if ($self->IsFlushKey(ord($char))) {
                  my $buf = $self->Buffer;
                  my $len = length($buf);
                  unless (($buf eq '') or ($self->IsFlushKey(ord(substr($buf, $len - 1))))) {
                     $self->Flush;
                  }
               }
               if ($call ne $self->LastCall) {
                  $self->Flush;
               }
               $self->LastCall($call);
               unless ($self->Buffer) { $self->BufStart($index); }
               $self->Buffer($self->Buffer . $char);
               if ($call eq 'overwrite') {
                  $self->ReplaceBuffer($self->ReplaceBuffer . $otxt);
               }
               return 1
            } else {
               carp "unexpected call '$call', it should have been 'write' or 'replace'\n";
            }
         }
      }
   }
   return 0
}

sub BufferMode {
   my $self = shift;
   if (@_) { $self->{BUFFERMODE} = shift; }
   return $self->{BUFFERMODE};
}

sub BufferModeBs {
   my $self = shift;
   return $self->BufferModeSwitch('backspace')
}

sub BufferModeDel {
   my $self = shift;
   return $self->BufferModeSwitch('delete')
}

sub BufferModeSwitch {
   my ($self, $call) = @_;
   $self->BufferNextCall(1);
   my $mode = $self->BufferMode;
   if ($mode ne $call) {
      if ($debug) { print "Switching to buffer mode '$call'\n"; }
      $self->Flush;
      $self->BufferMode($call);
   }
   return 0
}

sub BufferModeWr {
   my $self = shift;
   return $self->BufferModeSwitch('writeable')
}

sub BufferNextCall {
   my $self = shift;
   if (@_) { $self->{BUFFERNEXTCALL} = shift; }
   return $self->{BUFFERNEXTCALL};
}

sub BufStart {
   my $self = shift;
   if (@_) { $self->{BUFSTART} = shift; }
   return $self->{BUFSTART};
}

sub CancelCall {
   my $self = shift;
   $self->BufferNextCall(0);
   if ($debug) { print "received cancel \n"; }
   return 0;
}

sub CanUndo {
   my $self = shift;
   my $stack = $self->{UNDOSTACK};
   my $size = @$stack;
   return ($size > 0)
}

sub CanRedo {
   my $self = shift;
   my $stack = $self->{REDOSTACK};
   my $size = @$stack;
   return ($size > 0)
}

sub Clear {
   my $self = shift;
   $self->ResetRedo;
   $self->ResetUndo;
   $self->BufStart(0);
   $self->Buffer('');
   $self->BufferMode('');
   $self->ReplaceBuffer('');
   $self->LastCall('');
   return 0;
}

sub Flush {
   my $self = shift;
   my $buf = $self->Buffer;
   $self->Buffer('');
   my $bnc = $self->BufferNextCall;
   $self->BufferNextCall(0);
   my $pos = $self->BufStart;
   if ($buf ne '') {
      my $mode = $self->BufferMode;
      if ($mode eq 'delete') {
         $self->RecordUndo('remove', $pos, $buf, 0, $pos);
      } elsif ($mode eq 'backspace') {
         $self->RecordUndo('remove', $pos, $buf, 0);
      } elsif ($mode eq 'writeable') {
         my @call = ();
         my $ovr = $self->ReplaceBuffer;
         if ($ovr eq '') { push @call, 'write' } else { push @call, 'replace' }
         push @call, $pos, $buf;
         if ($ovr ne '') { push @call, $ovr }
         push @call, 0;
         if ($debug) { print "Flushing: @call'\n"; }
         $self->RecordUndo(@call);
      } else {
         carp "invalid buffer mode $mode\n";
      }
   }
   $self->BufferNextCall($bnc);
}

sub FlushKeys {
   my $self = shift;
   if (@_) { $self->{FLUSHKEYS} = shift; }
   return $self->{FLUSHKEYS};
}


sub IsFlushKey {
   my ($self, $key) = @_;
   my $fk = $self->FlushKeys;
   foreach (@$fk) {
      if ($key eq $_) {
         return 1
      }
   }
   return 0
}

sub LastCall {
   my $self = shift;
   if (@_) { $self->{LASTCALL} = shift; }
   return $self->{LASTCALL};
}

sub ReplaceBuffer {
   my $self = shift;
   if (@_) { $self->{REPLACEBUFFER} = shift; }
   return $self->{REPLACEBUFFER};
}

sub PullUndo {
   my $self = shift;
   my $stack = $self->{UNDOSTACK};
#   if ($debug) { print "undostack: ", Dumper $self->{UNDOSTACK} }
   return pop(@$stack);
}

sub PullRedo {
   my $self = shift;
   my $stack = $self->{REDOSTACK};
   return pop(@$stack);
}

sub PushUndo {
   my $self = shift;
   my $stack = $self->{UNDOSTACK};
   push(@$stack, @_);
#   if ($debug) { print "undostack: ", Dumper $self->{UNDOSTACK} }
}

sub PushRedo {
   my $self = shift;
   my $stack = $self->{REDOSTACK};
   push(@$stack, @_);
#   if ($debug) { print "redostack: ", Dumper $self->{UNDOSTACK} }
}

sub RecordReplace {
   my $self = shift;
   return $self->RecordUndo('replace', @_);
}

sub RecordRemove {
   my $self = shift;
   return $self->RecordUndo('remove', @_);
}

sub RecordUndo {
   my $self = shift;
   unless ($self->BufferCall(@_)) {
      $self->Flush;
      $self->ResetRedo;
      $self->PushUndo([@_]);
   }
   return 0;
}

sub RecordWrite {
   my $self = shift;
   return $self->RecordUndo('write', @_);
}

sub Redo {
   my $self = shift;
   $self->CancelCall;
   $self->Flush;
   if ($self->CanRedo) {
      my $o = $self->PullRedo;
      $self->PushUndo($o);
      my @call = @$o;
      my $mode = shift @call;
      if ($mode eq 'write') {
         return $self->Broadcast('dowrite', @call);
      } elsif ($mode eq 'remove') {
         return $self->Broadcast('doremove', @call);
      } elsif ($mode eq 'replace') {
         return $self->Broadcast('doreplace', @call)
      } else {
         carp "invalid redo mode $mode\n";
      }
   }
   return 0;
}

sub ResetRedo {
   my $self = shift;
   $self->{REDOSTACK} = [];
}

sub ResetUndo {
   my $self = shift;
   $self->{UNDOSTACK} = [];
}

sub Undo {
   my $self = shift;
   $self->CancelCall;
   $self->Flush;
   if ($self->CanUndo) {
      my $o = $self->PullUndo;
      $self->PushRedo($o);
      my @call = @$o;
      my $mode = shift @call;
      if ($mode eq 'write') {
         return $self->Broadcast('doremove', @call);
      } elsif ($mode eq 'remove') {
         return $self->Broadcast('dowrite', @call);
      } elsif ($mode eq 'replace') {
         my ($index, $old, $new, $sel, $ins) = @call;
         if ($self->Broadcast('doremove', $index, $new, 0)) {
            return $self->Broadcast('dowrite', $index, $old, $ins)
         }
      } else {
         carp "invalid undo mode $mode\n";
      }
   }
   return 0
}


1;
__END__


lib/Wx/Perl/PluggableTextCtrl/KeyEchoes.pm
package Wx::Perl::PluggableTextCtrl::KeyEchoes;

use strict;
use warnings;
use Carp;

use Wx qw( wxMOD_SHIFT );
use base qw( Wx::Perl::PluggableTextCtrl::BasePlugin );

my $debug = 1;

if ($debug) {
   use Data::Dumper;
}


sub new {
   my $class = shift;
   my $self = $class->SUPER::new(@_);

   $self->{ECHOES} = {};
   $self->Commands(
      'echo' => \&NewEcho,
      'key' => \&Key,
   );
   $self->Setup;

   return $self;
}

sub Echo {
   my $self = shift;
   my $key = shift;
   if (@_) { $self->{ECHOES}->{$key} = shift; }
   return $self->{ECHOES}->{$key};
}

sub EchoExists {
   my ($self, $key) = @_;
   return exists $self->{ECHOES}->{$key}
}


sub DefaultTable {
   my $self = shift;
   return [
        3 => 'copy',
        8 => \&DecodeBS,            #Backspace
        9 => \&DecodeTab,           #Tab
       13 => \&DecodeEnter,         #Enter or Ctrl-m, Ctrl-m does not produce a newline through the native handler. this might confuse plugins.
       22 => \&DecodePaste,         #Ctrl-v
       24 => \&DecodeCut,           #Ctrl-x
      127 => \&DecodeDel,           #Delete
      322 => 'insertkey',           #Insertkey pressed
   ]
}

sub DecodeBS {
   my $self = shift;
   unless ($self->Broadcast('backspace')) {
      my $tc = $self->TxtCtrl;
      if (($tc->GetInsertionPoint > 0) and $tc->IsEditable) {
         my ($selb, $sele) = $tc->GetSelection;
         if ($selb eq $sele) { #no selection
            my $p = $tc->GetInsertionPoint;
            if ($p > 0) {
               #broadcast remove: no selection => UndoRedo does it's own magic. Highlighter still needs to pick up on it.
               return $self->Broadcast('remove', $p - 1, $tc->GetRange($p -1, $p), 0, $p);
            }
         } else { #selection
            # broadcast uremove: there is a selection => UndoRedo records this straight into the undo stack.
            return $self->Broadcast('remove', $selb, $tc->GetStringSelection, 1, $tc->GetInsertionPoint);
         }
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0
}

sub DecodeCut {
   my $self = shift;
   unless ($self->Broadcast('cut')) {
      my $tc = $self->TxtCtrl;
      if ($tc->IsEditable) {
         my ($selb, $sele) = $tc->GetSelection;
         if ($selb ne $sele) { #there is a selection
            return $self->Broadcast('remove', $selb, $tc->GetStringSelection, 1, $tc->GetInsertionPoint);
         }
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0
}

sub DecodeDel {
   my $self = shift;
   unless ($self->Broadcast('delete')) {
      my $tc = $self->TxtCtrl;
      if (($tc->GetInsertionPoint < $tc->GetLastPosition) and $tc->IsEditable) {
         my ($selb, $sele) = $tc->GetSelection;
         if ($selb eq $sele) { #no selection
            my $p = $tc->GetInsertionPoint;
            if ($p < $tc->GetLastPosition) {
               return $self->Broadcast('remove', $p, $tc->GetRange($p, $p + 1), 0, $p);
            }
         } else {
            return $self->Broadcast('remove', $selb, $tc->GetStringSelection, 1, $tc->GetInsertionPoint);
         }
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0
}

sub DecodeEnter {
   my $self = shift;
   unless ($self->Broadcast('enter')) {
      my $tc = $self->TxtCtrl;
      if ($tc->IsMultiLine and $tc->IsEditable) {
         return $self->KeyWriteable(13)
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0
}

sub DecodePaste {
   my $self = shift;
   unless ($self->Broadcast('paste')) {
      my $tc = $self->TxtCtrl;
      if ($tc->IsEditable) {
         my ($selb, $sele) = $tc->GetSelection;
         my $ins = $tc->GetInsertionPoint;
         if ($selb eq $sele) { #there is no selection
            return $self->Broadcast('write', $ins, $tc->GetClipboardText, 0);
         } else {
            return $self->Broadcast('replace', $selb, $tc->GetStringSelection, $tc->GetClipboardText, 1);
         }
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0;
}

sub DecodeTab {
   my $self = shift;
   unless ($self->Broadcast('tab')) {
      my $tc = $self->TxtCtrl;
      if ($tc->IsEditable) {
         return $self->KeyWriteable(9)
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0
}

sub Key {
   my ($self, $event) = @_;
   my $k = $event->GetKeyCode;
   my $tc = $self->TxtCtrl;
   if ($debug) {
      my $r = $event->GetRawKeyCode;
      print "captured key: $k, raw: $r\n";
   }
   if ($tc->IsWriteable($k)) { #writeable character
      return $self->KeyWriteable($k);
   } elsif ($self->EchoExists($k)) {
      my $echo = $self->Echo($k);
      if (ref $echo) {
         return &$echo($self, $event)
      } else {
         if ($debug) { print "broadcasting $echo\n" }
         return $self->Broadcast($echo);
      }
   } 
   return 0;
}

sub KeyWriteable {
   my ($self, $key) = @_;
   unless ($self->Broadcast('writeable')) {
      my $tc = $self->TxtCtrl;
      if ($tc->IsEditable) {
         if ($tc->HasSelection) {
            my ($selb, $sele) = $tc->GetSelection;
            return $self->Broadcast('replace', $selb, $tc->GetStringSelection, chr($key), 1);
         } else {
            my $pos = $self->TxtCtrl->GetInsertionPoint;
            return $self->Broadcast('write', $pos, chr($key), 0);
         }
      } else {
         return $self->Broadcast('cancel')
      }
   }
   return 0;
}

sub NewEcho {
   my $self = shift;
   $self->Echo(@_);
   return 1
}

sub Setup {
   my ($self, $tb, $clr) = @_;
   unless (defined($tb)) {
      $tb = $self->DefaultTable;
   }
   if (defined($clr) and $clr) {
      $self->{ECHOS} = {};
   }
   my @table = @$tb;
   while (@table) {
      my $key = shift @table;
      my $echo = shift @table;
      $self->Echo($key, $echo);
   }
}


1;
__END__


lib/Wx/Perl/PluggableTextCtrl/BasePlugin.pm
package Wx::Perl::PluggableTextCtrl::BasePlugin;

use strict;
use warnings;
use Carp;

use vars qw($VERSION);
$VERSION="0.01";

my $debug = 0;

if ($debug) {
   use Data::Dumper;
}

sub new {
   my $proto = shift;
   my $class = ref($proto) || $proto;
   my $txtctrl = shift;
   my $self = {};
   $self->{TXTCTRL} = $txtctrl;
   $self->{COMMANDS} = {};

   bless ($self, $class);

   return $self;
}

sub Broadcast {
   my $self = shift;
   return $self->TxtCtrl->Broadcast(@_);
}

sub Command {
   my $self = shift;
   my $name = shift;
   if (@_) { $self->{COMMANDS}->{$name} = shift }
   return $self->{COMMANDS}->{$name}
}

sub Commands {
   my $self = shift;
   $self->{COMMANDS} = {};
   while (@_) {
      my $name = shift; my $call = shift;
      if ($debug) { print "setting command $name\n"; }
      $self->{COMMANDS}->{$name} = $call;
   }
}

sub Name {
   my $self = shift;
   my $name = ref $self;
   $name =~s/.*:://;
   if ($debug) { print "plugin name is $name\n" }
   return $name
}

sub Receive {
   my $self = shift;
   my $name = shift;
   if ($debug) { print "received $name\n"; print Dumper $self->{COMMANDS} }
   if (exists $self->{COMMANDS}->{$name}) {
      if ($debug) { print "executing $name\n" }
      my $cmd = $self->Command($name);
      return &$cmd($self, @_);
   }
   return 0
}

sub Require {
   my $self = shift;
   my $tc = $self->TxtCtrl;
   while (@_) {
      my $m = shift;
      if ($debug) { print "Requiring $m\n" }
      unless (defined($tc->FindPluginId($m))) {
         $tc->LoadPlugin($m);
      }
   }
}

sub TxtCtrl {
   my $self = shift;
   if (@_) { $self->{TXTCTRL} = shift; }
   return $self->{TXTCTRL};
}

1;
__END__

lib/Wx/Perl/PluggableTextCtrl.pm
package Wx::Perl::PluggableTextCtrl;

use strict;
use warnings;
use Carp;

use vars qw($VERSION);
$VERSION="0.01";

use Wx qw( :textctrl :font :colour );
use Wx::DND;
use Wx qw( wxTheClipboard );
use base qw( Wx::TextCtrl );
use Wx::Event qw( EVT_CHAR );

require Wx::Perl::PluggableTextCtrl::KeyEchoes;
require Wx::Perl::PluggableTextCtrl::UndoRedo;
require Wx::Perl::PluggableTextCtrl::Highlighter;

my $defaultfont = [10, wxFONTFAMILY_MODERN, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL, 0];

my $debug = 0;

if ($debug) {
   use Data::Dumper;
}

sub new {
   my $class = shift;
   my $self = $class->SUPER::new(@_);

   $self->{CALLBACK} = sub {};
   $self->{COMMANDS} = {
      'doremove' => \&DoRemove,
      'doreplace' => \&DoReplace,
      'dowrite' => \&DoWrite,
   };
   $self->{LISTENING} = 0;
   
   $self->{OVRMODE} = 0;
   $self->{PLUGINS} = [];

   $self->SetFont( Wx::Font->new(@$defaultfont) );
   EVT_CHAR($self, \&OnChar);

   return $self;
}

sub AppendText {
   my $self = shift;
   unless ($self->Broadcast('append', @_)) {
      $self->SUPER::Append(@_);
   }
}

sub Broadcast {
   my $self = shift;
   my $plugs = $self->{PLUGINS};
   my $flag = 0;
   foreach (@$plugs) {
      if ($_->Receive(@_)) {
         $flag = 1;
      }
   }
   return $flag;
}

sub Call2Remove {
   my ($self, $call, $index, $txt) = @_;
   if ($call =~ /.*remove$/) {
      return ($index, $index + length($txt))
   } else {
      carp "Call '$call' is not a remove type";
      return undef
   }
}

sub Call2Replace {
   my ($self, $call, $index, $old, $txt, $sel) = @_;
   if ($call =~ /.*replace$/) {
      return ($index, $index + length($old), $txt)
   } else {
      carp "Call '$call' is not a replace type";
      return undef
   }
}

sub Call2WriteText {
   my ($self, $call, $index, $txt) = @_;
   if ($call =~ /.*write$/) {
      return $txt
   } else {
      carp "Call '$call' is not a write type";
      return undef
   }
}

sub Callback {
   my $self = shift;
   if (@_) { $self->{CALLBACK} = shift; }
   return $self->{CALLBACK};
}

sub CanUndo {
   my $self = shift;
   return $self->Broadcast('canundo');
}

sub CanRedo {
   my $self = shift;
   return $self->Broadcast('canredo');
}

sub Clear {
   my $self = shift;
   unless ($self->Broadcast('clear')) {
      $self->SUPER::Clear;
   }
}

sub ClearSelection {
   my $self = shift;
   my $ins = $self->GetInsertionPoint;
   $self->SetSelection($ins, $ins);
}

sub Command {
   my $self = shift;
   my $name = shift;
   if (@_) { $self->{COMMANDS}->{$name} = shift }
   return $self->{COMMANDS}->{$name}
}

sub Copy {
   my $self = shift;
   unless ($self->Broadcast('copy')) {
      $self->SUPER::Copy;
   }
}

sub Cut {
   my $self = shift;
   unless ($self->Broadcast('cut')) {
      $self->SUPER::Cut;
   }
}

sub DoRemove {
   my $self = shift;
   my ($index, $txt, $sel, $ins) = @_;
   $self->ClearSelection;
   $self->SUPER::Remove($index, $index + length($txt));
   if (defined($ins)) {
      $self->SetInsertionPoint($ins);
   }
   return 1
}

sub DoReplace {
   my ($self, $index, $old, $txt, $sel, $ins) = @_;
   $self->ClearSelection;
   $self->SUPER::Replace($index, $index + length($old), $txt);
   if ($sel) {
      $self->SetSelection($index, $index + length($txt));
   }
   if (defined($ins)) {
      $self->SetInsertionPoint($ins);
   }
   return 1
}

sub DoWrite {
   my ($self, $index, $txt, $sel, $ins) = @_;
   $self->ClearSelection;
   $self->SetInsertionPoint($index);
   $self->SUPER::WriteText($txt);
   if ($sel) {
      $self->SetSelection($index, $index + length($txt));
   }
   if (defined($ins)) {
      $self->SetInsertionPoint($ins);
   }
   return 1
}

sub FindPluginId {
   my ($self, $name) = @_;
   my $plgs = $self->{PLUGINS};
   my $index = 0;
   foreach (@$plgs) {
      if ($name eq $plgs->[$index]->Name) {
         return $index
      }
      $index ++;
   }
   carp "Plugin $name is not loaded\n";
   return undef;
}

sub GetClipboardText {
   my $self = shift;
   my $txt = undef;
   if (wxTheClipboard->Open) {
      if ($debug) { print "Clipboard open\n" }
      my $textdata = Wx::TextDataObject->new;
      my $ok = wxTheClipboard->GetData( $textdata );
      if( $ok ) {
         $txt = $textdata->GetText;
      }
      if ($debug and defined($txt)) { print "Clipboard text: $txt\n" }
      wxTheClipboard->Close;
   }
   return $txt;
}

sub GetLineNumber {
   my ($self, $index) = @_;
   unless (defined($index)) { $index = $self->GetInsertionPoint };
   my ($col, $line) = $self->PositionToXY($index);
   return $line;
}

sub HasSelection {
   my $self = shift;
   my ($selb, $sele) = $self->GetSelection;
   return ($selb ne $sele)
}

# TODO make this unicode compatible
sub IsWriteable {
   my ($self, $key) = @_;
   if ((($key >= 32) and ($key < 127)) or (($key > 127) and ($key < 256))) {
      return 1
   }
   return 0
}

sub Listening {
   my $self = shift;
   if (@_) {
      my $new = shift;
      unless ($new eq $self->{LISTENING}) {
         my $plgs = $self->{PLUGINS};
         if ($new) {
            unshift @$plgs, $self
         } else {
            shift @$plgs
         }
         $self->{LISTENING} = $new
      }
   }
   return $self->{LISTENING}
}

sub LoadFile {
   my $self = shift;
   unless ($self->Broadcast('load', @_)) {
      $self->SUPER::LoadFile(@_);
   }
}

sub LoadPlugin {
   my $self = shift;
   while (@_) {
      my $plug = undef;
      my $name = shift;
      #Does anybody have a better idea for this?
      $name = "Wx::Perl::PluggableTextCtrl::$name";
      $plug = $name->new($self);
#      eval "require Wx::Perl::PluggableTextCtrl::$name; \$plug = Wx::Perl::PluggableTextCtrl::$name" . "->new(\$self);";
      if (defined($plug)) {
         $self->RegisterPlugin($plug);
      } else {
         carp "unable to load plugin $name\n";
      }
   }
}

sub Name {
   my $self = shift;
   my $name = ref $self;
   $name =~s/.*:://;
   if ($debug) { print "plugin name is $name\n" }
   return $name
}

sub OnChar {
   my ($self, $event) = @_;
   my $k = $event->GetKeyCode;
   if ($k eq 322) { #Insert key pressed, record flip insert/ovr mode.
      if ($self->OvrMode) {
         $self->OvrMode(0)
      } else {
         $self->OvrMode(1)
      }
   }
   unless ($self->Broadcast('key', $event)) {
      $event->Skip;
   }
   my $callback = $self->Callback;
   &$callback;
}

sub OvrMode {
   my $self = shift;
   if (@_) { $self->{OVRMODE} = shift; }
   return $self->{OVRMODE};
}

sub Paste {
   my $self = shift;
   unless ($self->Broadcast('paste')) {
      $self->SUPER::Paste;
   }
}

sub Plugin {
   my $self = shift;
   my $id = shift;
   my $plgs = $self->{PLUGINS};
   unless ($id =~ /^\d+$/) {
      $id = $self->FindPluginId($id);
   }
   if (@_) { 
      $self->{PLUGINS}->[$id] = shift; 
   }
   return $self->{PLUGINS}->[$id];
}

sub Receive {
   my $self = shift;
   my $name = shift;
#    if ($debug) { print "received $name\n"; print Dumper $self->{COMMANDS} }
   if (exists $self->{COMMANDS}->{$name}) {
      if ($debug) { print "executing $name\n" }
      my $cmd = $self->Command($name);
      return &$cmd($self, @_);
   }
   return 0
}

sub Redo {
   my $self = shift;
   unless ($self->Broadcast('redo')) {
      $self->SUPER::Redo;
   }
}

sub RegisterPlugin {
   my ($self, $plug) = @_;
   my $pl = $self->{PLUGINS};
   push @$pl, $plug;
}

sub Remove {
   my $self = shift;
   my @call = $self->Remove2Call(@_);
   unless ($self->Broadcast(@call)) {
      $self->SUPER::Remove(@_);
   }
}

sub Remove2Call {
   my ($self, $begin, $end) = @_;
   my $sel = 0;
   my ($selb, $sele) = $self->GetSelection;
   if (($selb eq $begin) and ($sele eq $end)) { $sel = 1 }
   return ('remove', $begin, $self->GetRange($begin, $end), $sel)
}

sub Replace {
   my $self = shift;
   my @call = $self->Replace2Call(@_);
   unless ($self->Broadcast(@call)) {
      $self->SUPER::Replace(@_);
   }
}

sub Replace2Call {
   my ($self, $begin, $end, $txt) = @_;
   my $sel = 0;
   my ($selb, $sele) = $self->GetSelection;
   if (($selb eq $begin) and ($sele eq $end)) { $sel = 1 }
   return ('replace', $begin, $txt, $self->GetRange($begin, $end), $sel)
}

sub SaveFile {
   my $self = shift;
   unless ($self->Broadcast('replace', @_)) {
      $self->SUPER::Replace(@_);
   }
}

sub NativePlugins {
   my $self = shift;
   return qw[ Highlighter KeyEchoes UndoRedo   ]
}

sub Syntax {
   my $self = shift;
   return $self->Broadcast('syntax', @_);
}

sub Undo {
   my $self = shift;
   unless ($self->Broadcast('undo')) {
      $self->SUPER::Undo;
   }
}

sub WriteText {
   my $self = shift;
   my @call = $self->WriteText2Call(@_);
   unless ($self->Broadcast(@call)) {
      $self->SUPER::WriteText(@_);
   }
}

sub WriteText2Call {
   my ($self, $txt) = @_;
   return ('write', $self->GetInsertionPoint, $txt, 0);
}


1;
__END__
